### PR TITLE
Fetch assets with a web worker

### PR DESCRIPTION
**Pull request by mzgoddard**
### Resolves

Load assets faster.

### Proposed Changes

When it makes sense and is possible load assets in a web worker. As well use fetch instead of nets xhr wrapper when possible. If we can't fetch in a worker or fetch in the main thread, fallback to the nets wrapper.

### Reason for Changes

Loading assets in a web worker shifts asset loading overhead, switching between native and js. It also reduces the chance of the js blocking work happening on one of those assets loading. These behaviour changes greatly reduce the time between operating on loading the raw data of assets.

### Benchmark Data

Can be compared against the baseline in https://github.com/LLK/scratch-vm/pull/1944.

project id | platform | run 1 | run 2 | run 3 | run 4 | run 5 | avg
--------- | ------ | ---- | ---- | ---- | ---- | ---- | ----
169401431 | chrome | 7124 | 6821 | 7042 | 6598 | 6794 | 6875.8
169401431 | firefox | 5946 | 4139 | 3739 | 4098 | 4456 | 4475.6
169401431 | safari | 10614 | 7203 | 7190 | 7281 | 6974 | 7852.4
169401431 | book | 18692 | 16970 | 17106 | 17232 | 16304 | 17260.8
173918262 | chrome | 1860 | 1472 | 1325 | 1283 | 1178 | 1423.6
173918262 | firefox | 1312 | 1188 | 1184 | 1255 | 1155 | 1218.8
173918262 | safari | 2207 | 1030 | 1155 | 1061 | 1079 | 1306.4
173918262 | book | 5642 | 3054 | 2489 | 2527 | 2921 | 3326.6
173918262 | pi | 26373 | 17820 | 16502 | 18263 | 16711 | 19133.8

### CODE REVIEW

File: src/BuiltinHelper.js [hunk 1]
```diff
@@ -152,9 +152,14 @@ class BuiltinHelper extends Helper {
      * Fetch an asset but don't process dependencies.
      * @param {AssetType} assetType - The type of asset to fetch.
      * @param {string} assetId - The ID of the asset to fetch: a project ID, MD5, etc.
-     * @return {Promise.<Asset>} A promise for the contents of the asset.
+     * @return {!Promise.<Asset>} A promise for the contents of the asset.
```
Comment (cwillisf) [inline src/BuiltinHelper.js]: The `!` means "never null" -- did you mean to add `?`?
(I think `?` is optional for reference types -- implied unless `!` is present -- but I'm a little fuzzy on that.)
See also: http://usejsdoc.org/tags-type.html

```suggestion
     * @return {?Promise.<Asset>} A promise for the contents of the asset.
```

File: src/FetchWorkerTool.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+        /**
+         * A map of ids to fetch job objects.
+         * @type {object}
+         */
+        this.jobs = {};
+
+        try {
+            if (this.getSupported) {
+                /* eslint global-require:0 */
```
Comment (cwillisf) [inline src/FetchWorkerTool.js]:
```suggestion
                // eslint-disable-next-line global-require
```

**Review by cwillisf — APPROVED** [review]
👍
I marked a couple minor suggestions but overall it looks reasonable to me.

One question: is there a reason to depend on `fetch` instead of using `nets` in the worker? Using `nets` in the worker would require a little bit more Webpack config but might simplify some of the code: the "tool" abstraction might become unnecessary for example. I don't feel strongly about it, though... just curious.

**Comment by mzgoddard** [comment]
- Commit @cwillisf's suggestions
- Squash fixups

**Comment by mzgoddard** [comment]
- Add benchmark data

File: src/ScratchStorage.js [hunk 3]
```diff
@@ -177,49 +177,56 @@ class ScratchStorage {
...
+            }
+
+            // Nothing went wrong but we couldn't find the asset.
+            return Promise.resolve(null);
+        };
+
+        return tryNextHelper()
+            .then(asset => {
+                if (asset === null) {
+                    return tryNextHelper();
```
Comment (kchadha) [inline src/ScratchStorage.js]: Why is this necessary when `tryNextHelper` already tries to handle null?

**Review by kchadha — COMMENTED** [review]
Had a question about the code in the comments, otherwise looks good to me

File: src/ScratchStorage.js (hunk 3 above)
Comment (mzgoddard) [reply src/ScratchStorage.js]: You're right its not needed. Its part of [this](https://github.com/LLK/scratch-storage/pull/78/files#diff-ba10d9fbbe07eb4ce18d86eda3f7ca2eL191) original handler mostly unmodified. I'll remove it. (It might also cause an infinite loop?)

Comment (mzgoddard) [reply src/ScratchStorage.js]: To clarify it'll loop when it can't find anything to load.

**Comment by mzgoddard** [comment]
- Remove the bit of code @kchadha pointed out

**Review by kchadha — APPROVED** [review]
LGTM

File: src/FetchWorkerTool.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+     * Request data from a server with a worker using fetch.
+     * @param {{url:string}} reqConfig - Request configuration for data to get.
+     * @param {{method:string}} options - Additional options to configure fetch.
+     * @returns {Promise.<Buffer>} Resolve to Buffer of data from server.
+     */
+    get ({url}, options = {method: 'GET'}) {
+        // TODO: This is rather knowledgeable of what data urls look like. Can
+        // it be abstracted or less knowledgable while letting FetchTool do the
+        // job with data?
+        if (/\/\w+\/get\/$|\.json/.test(url)) {
```
Comment (rschamp) [inline src/FetchWorkerTool.js]: What is the behavior of this if e.g., someone is using the scratch gui with different project/asset servers configured with a different URL structure? I think it's a high priority to avoid having this kind of deep knowledge if it's possible.

Comment (mzgoddard) [reply src/FetchWorkerTool.js]: I'm going to update WebHelper to choose a tool based on assetType (one for Projects and one for everything else).

**Comment by mzgoddard** [comment]
@rschamp @kchadha I changed WebHelper to have two tools, assetTool and projectTool. Passing a filter enum to ProxyTool, projectTool will use tools that are ready (and don't need to initialize like say a webworker). Let me know what you think.

File: src/FetchWorkerTool.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Is get supported?
+     *
+     * false if the environment does not workers, fetch, or fetch from inside a
+     * worker. Finding out the worker supports fetch is asynchronous and will
+     * guess that it does if the window does until the worker can inform us.
+     * @returns {boolean} Is get supported?
+     */
+    get getSupported () {
```
Comment (rschamp) [inline src/FetchWorkerTool.js]: This is confusing because it looks like a function name. Maybe `isGetSupported`?

File: src/FetchWorkerTool.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            };
+        })
+            .then(body => new Uint8Array(body));
+    }
+
+    /**
+     * Is sending supported? always false for FetchWorkerTool.
+     * @returns {boolean} Is sending supported?
+     */
+    get sendSupported () {
```
Comment (rschamp) [inline src/FetchWorkerTool.js]: Maybe `isSendSupported`?

File: src/FetchWorkerTool.worker.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * Register a step function.
+ *
+ * Step checks if there are completed jobs and if there are sends them to the
+ * parent. Then it checks the jobs count. If there are no further jobs, clear
+ * the step.
+ */
+const registerStep = function () {
+    intervalId = setInterval(() => {
+        if (complete.length) {
+            postMessage(complete.slice(), complete.map(response => response.buffer).filter(Boolean));
```
Comment (rschamp) [inline src/FetchWorkerTool.worker.js]: Could you please add a comment explaining the deal with transferring the buffer to the parent here?

File: src/ProxyTool.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+     * @returns {Promise.<Buffer>} Resolve to Buffer of data from server.
+     */
+    get (reqConfig, options) {
+        let toolIndex = 0;
+        const nextTool = err => {
+            const tool = this.tools[toolIndex++];
+            if (!tool.getSupported) {
+                return nextTool(err);
+            }
+            if (!tool) {
```
Comment (rschamp) [inline src/ProxyTool.js:72]: Shouldn't we check this existence before the `getSupported` property?

File: src/NetsTool.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * Send data to a server with nets.
+     * @param {{url:string}} reqConfig - Request configuration for data to send.
+     * @param {*} data - Data to send.
+     * @param {string} method - HTTP method to sending the data as.
+     * @returns {Promise.<Buffer|string|object>} Server returned metadata.
+     */
+    send (reqConfig, data, method) {
+        return new Promise((resolve, reject) => {
+            /* eslint global-require:0 */
```
Comment (rschamp) [inline src/NetsTool.js]: Can we use `eslint-disable-next-line` here instead?

**Review by rschamp — CHANGES_REQUESTED** [review]
@kchadha and I looked this over and think it looks good overall. We just have a few nits for readability/commenting.  Thanks!

File: src/ProxyTool.js (hunk 8 above)
Comment (mzgoddard) [reply src/ProxyTool.js:72]: I don't know how that happened. 😕

**Comment by mzgoddard** [comment]
- Made requested changes.

**Review by rschamp — APPROVED** [review]
Thank you @mzgoddard!
